mitxgraders/helpers/calc/expressions.py

```diff
@@ -575,90 +472,101 @@ def parse_algebra(self):
+        Try to parse a string and cache the result. ALWAYS clears storage.
         """
-        def handle_node(node):
-            """
-            Return the result representing the node, using recursion.
-
-            Call the appropriate action from self.actions for this node. As its inputs,
-            feed it the output of `handle_node` for each child node.
-            """
-            if not isinstance(node, ParseResults):
-                # Entry is either a (python) number or a string.
-                # Return it directly to the next level up.
-                return cast_np_numeric_as_builtin(node)
+        try:
+            BracketValidator.validate(expression)
+            tree = self.grammar.parseString(expression)[0]
+            parsed = MathExpression(expression,
+                                    tree,
+                                    self.variables_used,
+                                    self.functions_used,
+                                    self.suffixes_used)
+        except:
+            raise
+        finally:
+            self.reset_storage()
 
-            node_name = node.getName()
-            if node_name not in self.actions:  # pragma: no cover
-                raise Exception(u"Unknown branch name '{}'".format(node_name))
+        return parsed
 
-            action = self.actions[node_name]
-            handled_kids = [handle_node(k) for k in node]
+    def parse(self, expression):
+        """
+        If expression is in parser cache, return cached result, otherwise
+        delegate to raw_parse.
+        """
+        expression_no_whitespace = expression.replace(' ', '')
+        cache_key = expression_no_whitespace
+        if expression_no_whitespace in self.cache:
+            return self.cache[cache_key]
 
-            # Check for nan
-            if any(np.isnan(item) for item in handled_kids if isinstance(item, float)):
-                return float('nan')
+        try:
+            parsed = self.raw_parse(expression)
+        except ParseException:
+            msg = "Invalid Input: Could not parse '{}' as a formula"
+            raise UnableToParse(msg.format(expression))
 
-            # Compute the result of this node
-            result = action(handled_kids)
+        self.cache[cache_key] = parsed
+        return parsed
 
-            # All actions convert the input to a number, array, or list.
-            # (Only self.actions['arguments'] returns a list.)
-            as_list = result if isinstance(result, list) else [result]
+EvalMetaData = namedtuple('EvalMetaData', ['variables_used', 'functions_used', 'suffixes_used', 'max_array_dim_used'])
+class MathExpression(object):
+    """
+    Holds the parse tree for mathematical expression; returned by MathParser.
+
+    Attributes:
+        - expression (str): the original string that generated this parse tree
+        - variables_used (set)
+        - functions_used (set)
+        - suffixes_used (set)
+
```

Comment:
Is there a reason not to have `max_array_dim_used` here too? Maybe because we don't get that from just parsing; we compute it from an evaluation...

Reply:
Yes, that is why.

IMO, `max_array_dim_used` is weird as heck since it is based on evaluation but triggered by parsing into an ArrayNode. I mean, I do think that it behaves the way we want, I just think it's weird.

Consider:

```python
from mitxgraders.helpers.calc import evaluator, MathArray

A = MathArray([1, 2, 3])

result_1, meta_1 = evaluator('A', {'A': A})
print(meta_1.max_array_dim_used) # 0

result_1, meta_1 = evaluator('[A]', {'A': A})
print(meta_1.max_array_dim_used) # 2

```

Reply:
Ok, that example gave me a "Wait, what?" for a few seconds. I believe this is actually exactly the behavior that we want, though it is somewhat counter-intuitive. If you wanted `[A]` to have dimension one, that would actually be much harder for us to implement.